dbms/src/Interpreters/ActionsVisitor.cpp

```diff
@@ -107,37 +93,77 @@ SetPtr makeExplicitSet(
-    else if (left_tuple_depth + 1 == right_tuple_depth)
+
+    Block block;
+    auto col = left_arg_type->createColumn();
+    switch (right_arg_type->getTypeId())
     {
-        const auto * set_func = right_arg->as<ASTFunction>();
+        case TypeIndex::UInt8: [[fallthrough]];
+        case TypeIndex::UInt16: [[fallthrough]];
+        case TypeIndex::UInt32: [[fallthrough]];
+        case TypeIndex::UInt64: [[fallthrough]];
+        case TypeIndex::UInt128: [[fallthrough]];
+        case TypeIndex::Int8: [[fallthrough]];
+        case TypeIndex::Int16: [[fallthrough]];
+        case TypeIndex::Int32: [[fallthrough]];
+        case TypeIndex::Int64: [[fallthrough]];
+        case TypeIndex::Int128: [[fallthrough]];
+        case TypeIndex::Float32: [[fallthrough]];
+        case TypeIndex::Float64: [[fallthrough]];
+        case TypeIndex::Date: [[fallthrough]];
+        case TypeIndex::DateTime: [[fallthrough]];
+        case TypeIndex::String: [[fallthrough]];
+        case TypeIndex::FixedString: [[fallthrough]];
+        case TypeIndex::Enum8: [[fallthrough]];
+        case TypeIndex::Enum16: [[fallthrough]];
+        case TypeIndex::Decimal32: [[fallthrough]];
+        case TypeIndex::Decimal64: [[fallthrough]];
+        case TypeIndex::Decimal128: [[fallthrough]];
+        case TypeIndex::UUID:
+        {
+            col->insert(convertFieldToType(right_arg_value, *left_arg_type, right_arg_type.get()));
+            break;
+        }
+        // flatten compound values:
+        case TypeIndex::Array: [[fallthrough]];
+        case TypeIndex::Tuple: [[fallthrough]];
+        case TypeIndex::Set:
+        {
+            const Array & array = DB::get<const Array &>(right_arg_value);
+            if (array.size() == 0)
+                break;
 
-        if (!set_func || set_func->name != "tuple")
-            throw Exception("Incorrect type of 2nd argument for function " + node->name
-                            + ". Must be subquery or set of elements with type " + left_arg_type->getName() + ".",
+            for (size_t i = 0 ; i < array.size(); ++i)
+            {
+                col->insert(convertFieldToType(array[i], *left_arg_type, nullptr));
+            }
+            break;
+        }
+        default:
+            throw Exception("Unsupported value type at the right-side of IN:"
+                            + right_arg_type->getName() + ".",
                             ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
-
-        elements_ast = set_func->arguments;
     }
-    else
-        throw Exception("Invalid types for IN function: "
-                        + left_arg_type->getName() + " and " + right_arg_type->getName() + ".",
-                        ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+    block.insert(ColumnWithTypeAndName{std::move(col),
+                                       left_arg_type,
+                                       "dummy_" + left_arg_type->getName()});
```

Comment:
Why does it start with "dummy_"? It's not good but possible to tell user "Do not call your columns with `_` first. Special columns starts with it.". And it's not possible to tell user "Do not call your columns starting with `dummy_`". User would be confused finding `dummy_something` in error messages.